### PR TITLE
rsyslog: handle epipe transparently in syslog client

### PR DESCRIPTION
**Pull request by mhoffm-aiven**
If sending a message through syslog fails with EPIPE because the remote socket was closed we have to have a short retry mechanism. If we are not retrying the sending immediately we could continously have a situation where the socket timed out when we want to write something into it.

### CODE REVIEW

File: journalpump/rsyslog.py [hunk 1]
```diff
@@ -133,11 +149,30 @@ def close(self):
...
+                attempts += 1
+
+                if attempts > self.retry_attempts or not self._should_retry(ex=ex):
+                    raise
+
+                time.sleep(self.retry_backoff_seconds)
+
+    def _should_retry(self, *, ex):
+        if isinstance(ex, OSError):
+            return ex.errno is errno.EPIPE
```
Comment (rikonen) [inline journalpump/rsyslog.py:174]: `is` => `==`, otherwise this emits SyntaxWarning

**Review by rikonen — CHANGES_REQUESTED** [review]
Could also consider only catching `OSError` since we won't try retrying others but current code works just as well

File: journalpump/rsyslog.py (hunk 1 above)
Comment (mhoffm-aiven) [reply journalpump/rsyslog.py]: Thanks!

**Review by rikonen — DISMISSED** [review]

**Comment by safa-topal** [comment]
Nitpick: there's already some backoff retry implementation [here](https://github.com/aiven/journalpump/blob/c3b233f9b86f9af7a1b66e2d475f6954445c11ae/journalpump/senders/base.py#L115), would be great to reuse. One way of doing it: move the logic to a base class and inherit it from related senders.

**Comment by mhoffm-aiven** [comment]
> Nitpick: there's already some backoff retry implementation [here](https://github.com/aiven/journalpump/blob/c3b233f9b86f9af7a1b66e2d475f6954445c11ae/journalpump/senders/base.py#L115), would be great to reuse. One way of doing it: move the logic to a base class and inherit it from related senders.

The thing is, this is only supposed to do one quick retry to reconnect on EPIPE so having exponential backoff here does not feel appropriate.

**Comment by rikonen** [comment]
If we want to do something for this I'd change `retry_attempts: int` to `retry: bool` and `retry_backoff_seconds: float` to `wait_before_retry: float` to make it clear there's either one retry or not and we wait fixed amount of time before doing that retry.

File: journalpump/rsyslog.py [hunk 2]
```diff
@@ -133,11 +134,26 @@ def close(self):
...
+
+                self.socket.sendall(message[:self.max_msg - 1])
+                if len(message) >= self.max_msg:
+                    self.socket.sendall(b"\n")
+
+                break
+            except Exception as ex:  # pylint: disable=broad-except
+                self.close()
+
+                if not (retry and self._should_retry(ex=ex)):
```
Comment (rikonen) [inline journalpump/rsyslog.py:150]: Nit: imo `not retry or not self._should_retry(ex=ex)` would read easier but very subjective.

**Review by rikonen — APPROVED** [review]

**Review by safa-topal — APPROVED** [review]

**Pull request merged**